### PR TITLE
Slicing

### PR DESCRIPTION
**Pull request by saledouble**
Related to https://github.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/issues/108

### CODE REVIEW

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 1]
```diff
@@ -1043,133 +1051,118 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
...
-							CallSiteReference callSiteReference = callSites.next();
-
-							// get the (declared) called method at the call site.
-							MethodReference calledMethod = callSiteReference.getDeclaredTarget();
-
-							// is it a terminal operation? TODO: Should this be
-							// cached somehow? Collection of all terminal operation
-							// invocations?
-							if (isTerminalOperation(calledMethod)) {
-								// get the basic block for the call.
```
Comment (saledouble) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: move line 1048 - 1065 to the method below. The rest of the code in the block was not changed.

**Comment by saledouble** [comment]
I've tested 2 projects `threeten-extra` and `streamql`:


subject | previous running time(s) | current running time(s)
-- | -- | --
threeten-extra | 54.723 | 44.254
streamql | 849.381 | 785.035

**Comment by saledouble** [comment]
It's weird. My local JUnit testing was good and the output in console was different from the output travis.

**Comment by saledouble** [comment]
testNonInternalAPI7 cannot be passed on travis but it can pass on my local computer.

image

**Comment by khatchad** [comment]
Which test is failing on travis?

**Comment by khatchad** [comment]
Looks like https://travis-ci.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/builds/81931104 was the last passing test on master. HEAD isn't passing but I restarted it (looks like it timed out).

You need to also verify that you are receiving the same results in [this spreadsheet](https://docs.google.com/spreadsheets/d/1hLXshJv-efoZC9NQIOLqUEkKT4uiIIUF50wjAh_3kAM/edit#gid=872832558). I would start with smaller benchmarks first.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 2]
```diff
@@ -491,6 +494,8 @@ private static boolean signatureMatches(String[] operations, MethodReference met
 
 	private Set<InstanceKey> trackedInstances = new HashSet<>();
 
+	private static HashMap<CGNode, HashSet<CallSiteReference>> callSiteReferenceMap = new HashMap<>();
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: Purpose of this? Comments?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 3]
```diff
@@ -1291,4 +1284,63 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
...
+	 * 
+	 * @param cg:
+	 *            call graph
+	 * @return a pruned call graph
+	 */
+	public static Graph<CGNode> pruneCallGraph(final CallGraph cg) {
+		Graph<CGNode> partialGraph = GraphSlicer.prune(cg, new Predicate<CGNode>() {
+			@Override
+			public boolean test(CGNode node) {
+				return isStreamNode(node);
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: How is this related to https://github.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/issues/108#issue-274664134? Comments please.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 4]
```diff
@@ -1000,6 +1005,9 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
 		// for each rule.
 		SubMonitor ruleMonitor = subMonitor.split(70, SubMonitor.SUPPRESS_NONE).setWorkRemaining(ruleArray.length);
 
+		// slicing
+		Graph<CGNode> prunedCallGraph = pruneCallGraph(engine.getCallGraph());
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: Can we have a log on the size that we are saving? LIke number of nodes and edges before and after the refactoring?

**Comment by khatchad** [comment]
Also, are the tests passing on your local machine?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 5]
```diff
@@ -1291,4 +1284,63 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
...
+	 * @param node:
+	 *            CGNode
+	 * @return true/false
+	 */
+	public static boolean isStreamNode(CGNode node) {
+
+		boolean isStreamNode = false;
+
+		if (node.getMethod().getDeclaringClass().getClassLoader().getReference()
+				.equals(ClassLoaderReference.Application)) {
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: Not sure what application code has to do with this.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 6]
```diff
@@ -1291,4 +1284,63 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
...
+				// get the call site reference.
+				CallSiteReference callSiteReference = callSites.next();
+
+				// get the (declared) called method at the call site.
+				MethodReference calledMethod = callSiteReference.getDeclaredTarget();
+
+				// is it a terminal operation? TODO: Should this be
+				// cached somehow? Collection of all terminal operation
+				// invocations?
+				if (isTerminalOperation(calledMethod)) {
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: What do terminal operations have to do with this?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 7]
```diff
@@ -1291,4 +1284,63 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
...
+			@Override
+			public boolean test(CGNode node) {
+				return isStreamNode(node);
+			}
+		});
+		return partialGraph;
+	}
+
+	/**
+	 * Check whether a CGNode is reached by stream object
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: Where is the reaching definition analysis?

**Comment by saledouble** [comment]
>Also, are the tests passing on your local machine?

Yes.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java (hunk 4 above)
Comment (saledouble) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: Yes.

**Comment by saledouble** [comment]
>I don't see where you are querying whether the node is a "stream" node. I just see application and terminal operations, which has nothing to do with the bug that I can see.

If one of the called methods of a CGNode is the terminal operation, the node should be a `stream` node. Right?

**Comment by khatchad** [comment]
1. Is that consistent with "reaching definitions."
1. Is that general enough?

**Comment by khatchad** [comment]
I would suggest coming up with an algorithm and reviewing it before making any further changes.

**Comment by khatchad** [comment]
A first step may be to define "reaching definition."

**Comment by saledouble** [comment]
>"reaching definition."

The stream terminal operation is called in the method represented by CGNode?

**Comment by saledouble** [comment]
>I would suggest coming up with an algorithm and reviewing it before making any further changes.

Currently, I get all called methods of a CGNode and check whether at least one terminal operations exist. If they exist, the CGNode is a `stream` node and will not be pruned.

What's wrong with this algorithm?

**Comment by saledouble** [comment]
Currently, the code is failing to process return type and the parameters in the invoked method.

1. Return type

No method can get the return type.  I need to process manually:
```
3 = new <Primordial,Ljava/lang/String>@0
4 = getfield < Primordial, Ljava/lang/StringBuilder, value, <Primordial,[C> > 1
6 = getfield < Primordial, Ljava/lang/StringBuilder, count, <Primordial,I> > 1
invokespecial < Primordial, Ljava/lang/String, <init>([CII)V > 3,4,5,6 @13 exception:7
return 3
```

2. Parameters in the invoked methods

The following test is failed because of no terminal operations.

```
class A {
	@EntryPoint
	void m() {
		concat(new HashSet().parallelStream(),
				new HashSet().parallelStream()).count();
	}
}
```

**Comment by khatchad** [comment]
We already analyze return typed in the project.

Raffi Khatchadourian
Assistant Professor, Computer Science, Hunter College
Doctoral Faculty of the Graduate School and University Center's PhD Program in Computer Science
City University of New York
695 Park Avenue
Room HN 1090-H
New York, NY 10065
P: (212) 650-3988
F: (212) 772-5219
raffi.khatchadourian@hunter.cuny.edu
http://cuny.is/khatchad

On Aug 20, 2018 5:31 PM, Grace Tang <notifications@github.com> wrote:

Currently, the code is failing to process return type and the parameters in the invoked method.

  1.  Return type

No method can get the return type. I need to process manually:

3 = new <Primordial,Ljava/lang/String>@0
4 = getfield < Primordial, Ljava/lang/StringBuilder, value, <Primordial,[C> > 1
6 = getfield < Primordial, Ljava/lang/StringBuilder, count, <Primordial,I> > 1
invokespecial < Primordial, Ljava/lang/String, <init>([CII)V > 3,4,5,6 @13 exception:7
return 3


  1.  Parameters in the invoked methods

The following test is failed because of no terminal operations.

class A {
        @EntryPoint
        void m() {
                concat(new HashSet().parallelStream(),
                                new HashSet().parallelStream()).count();
        }
}

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 8]
```diff
@@ -1043,7 +1072,7 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
 				// TODO: Can this be somehow rewritten to get blocks corresponding to terminal
 				// operations?
 				// for each call graph node in the call graph.
-				for (CGNode cgNode : engine.getCallGraph())
+				for (CGNode cgNode : partialCallGraph)
```
[the remark this answers is not in the document]
Comment (saledouble) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: The problem is that the type of pruned call graph is Graph and the type of call graph is Callgraph. They are different. So I just changed this position first. If the pruned call graph works well, I will change it in everywhere.

**Comment by saledouble** [comment]
I found 2 issues:

1. Just checking BaseStream is not enough which may cause node missing.
2. Method `getPossibleTypes` could never return. `recursive call ` never stop.

**Comment by saledouble** [comment]
Casting is fixed.

**Comment by saledouble** [comment]
For solving the infinite recursive calls, I may add an array to record the checked instructions. If the instruction appears twice, then make the method return.

**Comment by saledouble** [comment]
Currently, the issue is that the test cases cannot produce the correct answer, but if I only test one which may produce the wrong results, it can pass.

**Comment by saledouble** [comment]
Eventually, pass! @khatchad Thank you!

**Comment by saledouble** [comment]
I've tested 4 subjects and the slicing code does not change the results.

subjects | previous running time(s) | current running time(s)
-- | -- | --
jdk8-experiments | 614.465 | 46.234
threeten-extra | 43.969 | 25.307
streamql | 822.628 | 33.9
monads | 15.426 | 3.282

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 9]
```diff
@@ -1170,6 +1172,7 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
 							}
 						}
 					}
+
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: Please remove this space.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 10]
```diff
@@ -1291,4 +1294,19 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
...
+	private static CallGraph pruneCallGraph(CallGraph callGraph, IClassHierarchy classHierarchy) {
+		LOGGER.info("The number of nodes in call graph: " + callGraph.getNumberOfNodes());
+		HashSet<CGNode> keep = new HashSet<>();
+		for (CGNode node: callGraph) {
+			if(Util.isStreamNode(node, classHierarchy))
+				keep.add(node);
+		}
+		
+		PrunedCallGraph prunedCallGraph = new PrunedCallGraph(callGraph, keep);
+		LOGGER.info("The number of nodes in partial graph: " + prunedCallGraph.getNumberOfNodes());
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: What were the differences you got here?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 11]
```diff
@@ -1291,4 +1294,19 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
...
+		
+		return prunedCallGraph;
+	}
+
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: Please remove this space.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 12]
```diff
@@ -407,13 +410,31 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
...
+		checkInstructions.clear();
+		return getPossibleTypes(valueNumber, inference);
+	}
+
 	static Collection<TypeAbstraction> getPossibleTypes(int valueNumber, TypeInference inference) {
 		Set<TypeAbstraction> ret = new HashSet<>();
-		Value value = inference.getIR().getSymbolTable().getValue(valueNumber);
+		Value value;
+		try {
+			value = inference.getIR().getSymbolTable().getValue(valueNumber);
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Why would you get an IllegalArgumentException here?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 13]
```diff
@@ -407,13 +410,31 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
...
+		} catch (IllegalArgumentException exception) {
+			return ret;
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Seems dangerous. We would at the very least need a log statement here.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java (hunk 12 above)
Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Also, the indentation here isn't correct.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 14]
```diff
@@ -407,13 +410,31 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
 		return lineNumberFromIR;
 	}
 
+	static private LinkedList<Value> checkInstructions = new LinkedList<>();
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Seems super dangerous. And, there's no documentation.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 15]
```diff
@@ -407,13 +410,31 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
...
+
+	static Collection<TypeAbstraction> getPossibleStreamTypes(int valueNumber, TypeInference inference) {
+		checkInstructions.clear();
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Very dangerous to use such variables.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 16]
```diff
@@ -407,13 +410,31 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
...
+			value = inference.getIR().getSymbolTable().getValue(valueNumber);
+		} catch (IllegalArgumentException exception) {
+			return ret;
+		}
 
 		// TODO: Should really be using a pointer analysis here rather than
 		// re-implementing one using PhiValue.
 		if (value instanceof PhiValue) {
+
+			if (checkInstructions.contains(value))
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: I'm not convinced that this is needed. Also, there's no comments as to why it is needed.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java (hunk 14 above)
Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: It's also slow. Can you use a HashSet?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 17]
```diff
@@ -761,4 +785,48 @@ private static boolean wouldOrderingBeConsistent(final Collection<TypeAbstractio
 
 	private Util() {
 	}
+
+	public static boolean isStreamNode(CGNode node, IClassHierarchy classHierarchy) {
+		if (checkDeclaredClass(node))
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: This method name should be a question.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 18]
```diff
@@ -761,4 +785,48 @@ private static boolean wouldOrderingBeConsistent(final Collection<TypeAbstractio
...
+					.mapToObj(i -> instruction.getUse(i)).flatMap(u -> getPossibleStreamTypes(u, inference).stream());
+
+			if (Stream.concat(defs, uses).anyMatch(t -> implementsBaseStream(t.getTypeReference(), classHierarchy)))
+				return true;
+		}
+
+		return false;
+	}
+
+	// check declared class for CGNode
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Javadoc please.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 19]
```diff
@@ -761,4 +785,48 @@ private static boolean wouldOrderingBeConsistent(final Collection<TypeAbstractio
...
+	private static boolean checkDeclaredClass(CGNode node) {
+		IMethod method = node.getMethod();
+		if (isBaseStream(method.getDeclaringClass()))
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Why not implementsBaseStream?

File: edu.cuny.hunter.streamrefactoring.tests/META-INF/MANIFEST.MF [hunk 20]
```diff
@@ -9,8 +9,7 @@ Fragment-Host: org.eclipse.jdt.ui
 Bundle-RequiredExecutionEnvironment: JavaSE-1.8
 Require-Bundle: org.junit,
  edu.cuny.citytech.refactoring.common.core;bundle-version="1.1.0",
- edu.cuny.citytech.refactoring.common.tests;bundle-version="1.5.0",
- org.eclipse.jdt.ui.tests.refactoring;bundle-version="3.13.100"
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.tests/META-INF/MANIFEST.MF]: Strange. I thought I already removed this in master.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java (hunk 16 above)
Comment (saledouble) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: If there is no this, the method could be recursively called again and again and will never return.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java (hunk 12 above)
Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: How do you know that?

Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: I think it's right but is there some documentation somewhere?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java (hunk 16 above)
Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Are you experiencing infinite recursion with this method?

Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: OK I tried removing it and it looks like it is happening: https://travis-ci.com/ponder-lab/Optimize-Java-8-Streams-Refactoring/builds/82575224?utm_source=email&utm_medium=notification

Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: So, it's not that there "could be," *there is*.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java (hunk 19 above)
Comment (saledouble) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Because the types of the parameters for `implementsBaseStream` should be ITypebinding or TypeReference, but the parameter type for isBaseStream is `IClass`, which means I can directly use it.

Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Use the `IClass` to get the `TypeReference`. I think you can use the `IClassHierarchy` to do so but don't quote me.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java [hunk 21]
```diff
@@ -1291,4 +1293,21 @@ private void propagateStreamInstanceProperty(Collection<InstanceKey> streamInsta
 		}
 		return ret;
 	}
+
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/StreamStateMachine.java]: Javadoc here, please. How is the pruning done?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 22]
```diff
@@ -407,13 +409,38 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
...
 
+	/**
+	 * This set is used to store visited values. Without this set, the method
+	 * getPossibleTypes could be infinitely recursively called and would never
+	 * return. So adding this set means to add a base case.
+	 */
+	static private HashSet<Value> seenValues = new HashSet();
+
+	static Collection<TypeAbstraction> getPossibleStreamTypes(int valueNumber, TypeInference inference) {
+		seenValues.clear();
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Why is the `clear()` only here?

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 23]
```diff
@@ -407,13 +409,38 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
...
+		return getPossibleTypes(valueNumber, inference);
+	}
+
 	static Collection<TypeAbstraction> getPossibleTypes(int valueNumber, TypeInference inference) {
 		Set<TypeAbstraction> ret = new HashSet<>();
-		Value value = inference.getIR().getSymbolTable().getValue(valueNumber);
+		Value value;
+		try {
+			value = inference.getIR().getSymbolTable().getValue(valueNumber);
+		} catch (IllegalArgumentException exception) {
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Again, please replace with conditional check.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java (hunk 22 above)
Comment (saledouble) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Each time call `getPossibleTypes`, this set must be cleared. The reason that infinite recursion happens here is that if the value is `PhiValue`, the `Uses` could be visited by the method before. 
For example:
No. 1 call:
**Value 3 -> Uses 5 7**
No. 2 call:
Value 5 -> return type
**Value 7** -> Uses 4 **3**
No. 3 call:
**Value 3 -> Uses 5 7**
Becase 3 is visited in the  No. 1 call, so the No. 3 call should be quited. That's why I added a set to store the visited values.

Comment (khatchad) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Why is it *only* here, meaning that there are other places that may need to be considered (like the other overloaded methods).

Comment (saledouble) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Done.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 24]
```diff
@@ -458,6 +485,7 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
 					// type is java.lang.Object.
 					// Find the return type of the instruction.
 					TypeInference inference = TypeInference.make(node.getIR(), false);
+					seenValues.clear();
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Why are you clearing it here? Comments? I was speaking of other *overloaded* methods that call this method.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 25]
```diff
@@ -761,4 +792,63 @@ private static boolean wouldOrderingBeConsistent(final Collection<TypeAbstractio
...
+
+	/**
+	 * Check declared class for CGNode
+	 * 
+	 * @param node:
+	 *            CGNode in the CallGraph
+	 */
+	private static boolean isDeclaredStreamClass(CGNode node) {
+		IMethod method = node.getMethod();
+		if (isBaseStream(method.getDeclaringClass()))
```
Comment (khatchad) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Still not correct. implementsBaseStream.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java (hunk 24 above)
Comment (saledouble) [reply edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: No overloaded method calls `getPossibleTypes()`. Besides the method that I created, this is the only place that `getPossibleTypes()` is called.

File: edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java [hunk 26]
```diff
@@ -458,7 +488,8 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
 					// type is java.lang.Object.
 					// Find the return type of the instruction.
 					TypeInference inference = TypeInference.make(node.getIR(), false);
-					Collection<TypeAbstraction> returnTypes = Util.getPossibleTypes(valueNumber, inference);
+					// Get all possible types
+					Collection<TypeAbstraction> returnTypes = Util.getAllPossibleTypes(valueNumber, inference);
```
Comment (saledouble) [inline edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java]: Using `getPossibleTypes()` is not safe. So I modified it to `getAllPossibleTypes()`
